src/translator/nth_element.cpp

```diff
@@ -59,23 +31,35 @@ class NthElementCPU {
     const auto inputN    = scores->shape()[-2];
     const auto dimBatch  = scores->shape()[-4];
     ABORT_IF(inputN != (isFirst ? 1 : N), "Input tensor has wrong beam dim??"); // @TODO: Remove isFirst argument altogether
-
-    std::vector<int> cumulativeBeamSizes(dimBatch + 1, 0);
-    std::vector<int> batchFirstElementIdxs(dimBatch + 1, 0);
-
-    for(int batchIdx = 0; batchIdx < dimBatch; ++batchIdx) {
-      cumulativeBeamSizes[batchIdx + 1] = (batchIdx + 1) * (int)N;
-      batchFirstElementIdxs[batchIdx + 1] += (batchIdx + 1) * inputN * vocabSize;
-      ABORT_IF(cumulativeBeamSizes[batchIdx + 1] != cumulativeBeamSizes[batchIdx] + (int)N, "cumulativeBeamSizes wrong??");
-      ABORT_IF((isFirst ? batchIdx + 1 : cumulativeBeamSizes[batchIdx + 1]) != (batchIdx + 1) * inputN, "inputN wrong??");
-    }
-    ABORT_IF(cumulativeBeamSizes.back() != dimBatch * N, "cumulativeBeamSizes.back() wrong??");
+    const float* scoresData = scores->data();
 
     size_t maxSize = N * dimBatch;
     h_res.resize(maxSize);
     h_res_idx.resize(maxSize);
+    size_t pos = 0; // iterates through h_res and h_res_idx
+
+    size_t batchOffset = inputN * vocabSize;
+    std::vector<int> idxs(batchOffset); // re-used for each batch
+    std::iota(idxs.begin(), idxs.end(), 0);
+
+    for(size_t batchIdx = 0; batchIdx < dimBatch; ++batchIdx) {
 
-    selectNBest(scores->data(), batchFirstElementIdxs, cumulativeBeamSizes);
+      std::partial_sort(
+        idxs.begin(),
+        idxs.begin() + N,
+        idxs.end(),
+        [&](int a, int b) { return scoresData[a] > scoresData[b]; }
+      );
+
+      for(int temp = 0; temp < N; ++temp) {
```

Comment:
I would suggest not using the name `temp` if it's just for iteration. `temp` suggests that it will be used to temporary store an intermediate value during some computation, but here we just have an int iterator.

Also, it can be size_t instead of int.